### PR TITLE
unordered categorical

### PR DESCRIPTION
**Pull request by sonichi**
rounding bug

### CODE REVIEW

File: flaml/searcher/flow2.py [hunk 1]
```diff
@@ -190,7 +190,8 @@ def _init_search(self):
         # record intermediate trial cost
         self._trial_cost = {}
         self._same = False  # whether the proposedd config is the same as best_config
-        self._init_phrase = True  # initial phase to increase initial stepsize
+        self._init_phase = True  # initial phase to increase initial stepsize
+        self._trunc = 0
```
Comment (qingyun-wu) [inline flaml/searcher/flow2.py:194]: Set it as None?
```suggestion
        self._trunc = None
```

File: flaml/searcher/flow2.py [hunk 2]
```diff
@@ -494,8 +501,10 @@ def on_trial_result(self, trial_id: str, result: Dict):
             # record the cost in case it is pruned and cost info is lost
             self._trial_cost[trial_id] = cost
 
-    def rand_vector_unit_sphere(self, dim) -> np.ndarray:
+    def rand_vector_unit_sphere(self, dim, trunc=1) -> np.ndarray:
         vec = self._random.normal(0, 1, dim)
+        if 0 < trunc < dim:
```
Comment (qingyun-wu) [inline flaml/searcher/flow2.py:506]:
```suggestion
        if trunc:
```

File: flaml/searcher/flow2.py [hunk 3]
```diff
@@ -494,8 +501,10 @@ def on_trial_result(self, trial_id: str, result: Dict):
...
+            vec[np.abs(vec).argsort()[:dim - trunc]] = 0
```
Comment (qingyun-wu) [inline flaml/searcher/flow2.py:507]:
```suggestion
            vec[np.abs(vec).argsort()[:max(dim - trunc, 0)]] = 0
```

File: flaml/searcher/flow2.py [hunk 4]
```diff
@@ -555,25 +565,46 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
...
+            # random
+            for i, key in enumerate(self._tunable_keys):
+                if self._direction_tried[i] != 0:
+                    for _, generated in generate_variants({'config': {
+                        key: self.space[key]
+                    }}):
+                        if generated['config'][key] != best_config[key]:
+                            config[key] = generated['config'][key]
+                            return unflatten_dict(config)
+                        break
```
Comment (qingyun-wu) [inline flaml/searcher/flow2.py:599]: why adding break here?

File: flaml/searcher/flow2.py [hunk 5]
```diff
@@ -555,25 +565,46 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
...
+        else:
+            # check if config == best_config
```
Comment (qingyun-wu) [inline flaml/searcher/flow2.py:601]: What is this for?

File: flaml/searcher/flow2.py [hunk 6]
```diff
@@ -494,8 +501,10 @@ def on_trial_result(self, trial_id: str, result: Dict):
             # record the cost in case it is pruned and cost info is lost
             self._trial_cost[trial_id] = cost
 
-    def rand_vector_unit_sphere(self, dim) -> np.ndarray:
+    def rand_vector_unit_sphere(self, dim, trunc=1) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/searcher/flow2.py:504]: I'll use 0

File: flaml/searcher/flow2.py (hunk 1 above)
Comment (sonichi) [reply flaml/searcher/flow2.py]: I added comments about self._trunc

File: flaml/searcher/flow2.py (hunk 4 above)
Comment (sonichi) [reply flaml/searcher/flow2.py:600]: same as before

File: flaml/searcher/flow2.py (hunk 5 above)
Comment (sonichi) [reply flaml/searcher/flow2.py:602]: If the rounded config is equal to the best config, then the incumbent would not be moved. By moving the incumbent, we increase the chance to escape that local point.

**Comment by sonichi** [comment]
close #125 #126

**Review by qingyun-wu — APPROVED** [review]

**Pull request closed without merging**